src/cluster/src/check.rs

```diff
@@ -109,42 +125,225 @@ pub enum CheckError {
     #[error("Minikube tunnel not found, retrying")]
     MinikubeTunnelNotFoundRetry,
 
-    // Default unhandled K8 client error
+    /// Default unhandled K8 client error
     #[error("Unhandled K8 client error")]
     UnhandledK8ClientError,
+
+    /// Kubectl not found
+    #[error("Kubectl not found")]
+    KubectlNotFoundError {
+        /// source : IoError
+        #[from]
+        source: IoError,
+    },
+
+    /// Unable to parse kubectl version
+    #[error("Unable to parse kubectl version")]
+    KubectlVersionError,
+
+    /// The installed version of Kubectl is incompatible
+    #[error("Must have kubectl version {required} or later. You have {installed}")]
+    IncompatibleKubectlVersion {
+        /// The currently-installed helm version
+        installed: String,
+        /// The minimum required helm version
+        required: String,
+    },
+
+    /// Check permissions to create k8 resources
+    #[error("Permissions to create {resource} denied")]
+    PermissionError {
+        /// Name of the resource
+        resource: String,
+    },
+
+    /// Error while fetching create permissions for a resource
+    #[error("Unable to fetch permissions")]
+    FetchPermissionError,
+
+    /// One or more pre flight checks have failed
+    #[error("Some pre-install checks have failed.")]
+    PreFlightCheckError,
 }
 
-/// Getting server hostname from K8 context
-pub fn check_cluster_server_host() -> Result<(), CheckError> {
-    let config = K8Config::load()?;
-    let context = match config {
-        K8Config::Pod(_) => return Ok(()),
-        K8Config::KubeConfig(context) => context,
-    };
+#[async_trait]
+pub trait InstallCheck: Send + Sync + 'static {
+    /// perform check, if successful return success message, if fail, return fail message
+    async fn perform_check(&self) -> Result<String, CheckError>;
```

Comment:
Based on previous comment, this should return structure which indicates
- problem
- where?
- remediation steps

And should not return Error unless it countered unexpected problem

Reply:
Do you think it will be better to extend the trait to include a `try_fix` method which will be called after each check and try to fix the error? Also we do not try to fix these errors in the pre flight checks , we just display these errors. I don't know if we we want to provide the fix while running the check command. So do you suggest if we should we include another flag here for autofix?

Reply:
I agree, having `try_fix` will be good.  This should not be invoked in the pre-flight.